hal/src/stm32f2xx/sleep_hal.cpp

```diff
@@ -175,17 +202,69 @@ static int validateNetworkWakeupSource(hal_sleep_mode_t mode, const hal_wakeup_s
     return SYSTEM_ERROR_NONE;
 }
 
+static int validateUsartWakeupSource(hal_sleep_mode_t mode, const hal_wakeup_source_usart_t* usart) {
+    if (!hal_usart_is_enabled(usart->serial)) {
+        return SYSTEM_ERROR_INVALID_STATE;
+    }
+    if (mode != HAL_SLEEP_MODE_STOP) {
+        return SYSTEM_ERROR_NOT_SUPPORTED;
+    }
+    return SYSTEM_ERROR_NONE;
+}
+
 static int validateWakeupSource(hal_sleep_mode_t mode, const hal_wakeup_source_base_t* base) {
     if (base->type == HAL_WAKEUP_SOURCE_TYPE_GPIO) {
         return validateGpioWakeupSource(mode, reinterpret_cast<const hal_wakeup_source_gpio_t*>(base));
     } else if (base->type == HAL_WAKEUP_SOURCE_TYPE_RTC) {
         return validateRtcWakeupSource(mode, reinterpret_cast<const hal_wakeup_source_rtc_t*>(base));
     } else if (base->type == HAL_WAKEUP_SOURCE_TYPE_NETWORK) {
         return validateNetworkWakeupSource(mode, reinterpret_cast<const hal_wakeup_source_network_t*>(base));
+    } else if (base->type == HAL_WAKEUP_SOURCE_TYPE_USART) {
+        return validateUsartWakeupSource(mode, reinterpret_cast<const hal_wakeup_source_usart_t*>(base));
     }
     return SYSTEM_ERROR_NOT_SUPPORTED;
 }
 
+static void enterSleepSleep(uint8_t highAccuracy) {
```

Comment:
See https://github.com/particle-iot/device-os/pull/2172/files#r475966911

Also, this implementation is different and uses HSE instead. Which one are we using?

Reply:
I will re-estimate the power consumption using HSI against using HSE. If the diversity is minimum, then let's use HSE instead, otherwise, we're going to using HSI for analog wakeup source and HSE for USART wakeup source. The reason we have to use HSE  for USART is that USART need accurate clock to guarantee the baudrate.

Reply:
Sounds good. I'd prefer that we used HSE as well if the difference is negligible. This will also allow to enter/exit sleep mode more quickly.

Reply:
For the USART wakeup source, we need to enable PLL so that the baudrate would not be affected, in which case it results in ~17mA average current. While using HSI only results in ~6mA average current. So we should use HSI for analog wakeup source conditionally, while use HSE along with PLL for USART wakeup source.

Reply:
Or use HSE for analog wakeup source as well, but need to disable PLL to optimize the power consumption.

Reply:
How much of a difference is HSE without PLL vs HSI?

Reply:
It's ~1mA higher than HSI if using HSE only.